File: src/commands.zig
const std = @import("std");
const http = std.http;

pub const Colors = struct {
    pub const red = "\x1b[31m";
    pub const white = "\x1b[37m";
    pub const yellow = "\x1b[33m";
    pub const green = "\x1b[32m";
};

fn printFmt(comptime fmt: []const u8, args: anytype) !void {
    const allocator = std.heap.page_allocator;
    const text = try std.fmt.allocPrint(allocator, fmt, args);
    defer allocator.free(text);
    try std.fs.File.stdout().writeAll(text);
}

fn repeatByte(allocator: std.mem.Allocator, byte: u8, count: usize) ![]u8 {
    const buffer = try allocator.alloc(u8, count);
    @memset(buffer, byte);
    return buffer;
}

pub fn displayHelpData(alloc: std.mem.Allocator) !void {
    const message = "Welcome to Zippy!";
    const boxWidth: usize = message.len + 4;
    const dash_count = boxWidth - 2;

    const lineBytes = try alloc.alloc(u8, dash_count * 3);
    defer alloc.free(lineBytes);
    for (0..dash_count) |i| {
        const offset = i * 3;
        lineBytes[offset + 0] = 0xE2;
        lineBytes[offset + 1] = 0x94;
        lineBytes[offset + 2] = 0x80;
    }
    const line = lineBytes;

    const padding_count = (boxWidth - message.len - 1) / 2;
    const padding = try repeatByte(alloc, ' ', padding_count);
    defer alloc.free(padding);

    try printFmt("{s}┌{s}┐\n", .{ Colors.red, line });
    try printFmt("{s}│{s}{s}{s}{s}{s}│\n", .{ Colors.red, padding, Colors.green, message, Colors.red, padding });
    try printFmt("{s}└{s}┘\n", .{ Colors.red, line });

    try printFmt("{s}Example:\n", .{Colors.green});
    try printFmt("{s}  zippy app/Main.hs\n\n", .{Colors.white});

    try printFmt("{s}Commands:\n", .{Colors.yellow});
    try printFmt("{s}  --help      Display help information\n", .{Colors.white});
    try printFmt("{s}  --version   Display version information/Check for updates\n", .{Colors.white});
    try printFmt("{s}  --config    Configure Zippy\n", .{Colors.white});
    try printFmt("{s}  --log       Display Zippy log\n", .{Colors.white});
    try printFmt("{s}  --clear     Clear Zippy log\n", .{Colors.white});
    try printFmt("{s}  --credits   Display credits\n", .{Colors.white});
}

pub fn displayConfigData() !void {
    try printFmt("{s}Configuration options (Zippy.json):\n", .{Colors.yellow});
    try printFmt("{s}  delay (u64 microseconds)        Debounce between checks (default 1000000)\n", .{Colors.white});
    try printFmt("{s}  cmd (string)                    Command to run; placeholders {file}, {dir}\n", .{Colors.white});
    try printFmt("{s}  save_log (bool)                Enable log file (default false)\n", .{Colors.white});
    try printFmt("{s}  log_path (string)              Path to log file (default zippy.log)\n", .{Colors.white});
}

pub fn displayLogData() !void {
    try printFmt("{s}--log shows log file contents when save_log=true.\n", .{Colors.yellow});
    try printFmt("{s}Log file path is configured via log_path in Zippy.json.\n", .{Colors.white});
    try printFmt("{s}When save_log=true, child command output is also tee'd to the log.\n", .{Colors.white});
}

pub fn displayClearData() !void {
    try printFmt("{s}--clear truncates the configured log file when save_log=true.\n", .{Colors.yellow});
}

pub fn displayCreditsData() !void {
    try printFmt("{s}Credits:\n", .{Colors.yellow});
    try printFmt("{s}Developed by: Ewen MacCulloch\n", .{Colors.green});
    try printFmt("{s}GitHub: Voyrox\n", .{Colors.green});
    try printFmt("{s}\n", .{Colors.white});
}

pub fn displayVersionData() !void {
    const currentVersion = "v1.3.1";

    try printFmt("{s}  Current version: {s}{s}\n", .{ Colors.green, currentVersion, Colors.white });
    try printFmt("{s}  https://github.com/Voyrox/Zippy/releases/latest\n", .{Colors.white});
}

File: src/commands.cpp
#include "commands.hpp"

#include <iostream>
#include <string>

namespace
{

    struct Colors
    {
        static constexpr const char *red = "\x1b[31m";
        static constexpr const char *white = "\x1b[37m";
        static constexpr const char *yellow = "\x1b[33m";
        static constexpr const char *green = "\x1b[32m";
        static constexpr const char *reset = "\x1b[0m";
    };

    void printLine(const std::string &text)
    {
        std::cout << text;
    }

    std::string repeatChar(char value, std::size_t count)
    {
        return std::string(count, value);
    }

}

namespace commands
{

    void displayHelpData()
    {
        const std::string message = "Welcome to Zippy!";
        const std::size_t boxWidth = message.size() + 4;
        const std::string line = repeatChar('-', boxWidth - 2);
        const std::size_t paddingCount = (boxWidth - message.size() - 2) / 2;
        const std::string padding = repeatChar(' ', paddingCount);

        printLine(std::string(Colors::red) + "+" + line + "+\n" + Colors::reset);
        printLine(std::string(Colors::red) + "|" + Colors::reset + padding + Colors::green + message + Colors::reset + padding + " |\n");
        printLine(std::string(Colors::red) + "+" + line + "+\n" + Colors::reset);
        printLine(std::string(Colors::green) + "Example:\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  zippy app/Main.hs\n\n" + Colors::reset);
        printLine(std::string(Colors::yellow) + "Commands:\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  --help      Display help information\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  --version   Display version information/Check for updates\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  --config    Configure Zippy\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  --log       Display Zippy log\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  --clear     Clear Zippy log\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  --credits   Display credits\n" + Colors::reset);
    }

    void displayConfigData()
    {
        printLine(std::string(Colors::yellow) + "Configuration options (Zippy.json):\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  delay (u64 microseconds)        Debounce between checks (default 1000000)\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  cmd (string)                    Command to run; placeholders {file}, {dir}\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  save_log (bool)                 Enable log file (default false)\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  log_path (string)               Path to log file (default zippy.log)\n" + Colors::reset);
    }

    void displayLogData()
    {
        printLine(std::string(Colors::yellow) + "--log shows log file contents when save_log=true.\n" + Colors::reset);
        printLine(std::string(Colors::white) + "Log file path is configured via log_path in Zippy.json.\n" + Colors::reset);
        printLine(std::string(Colors::white) + "When save_log=true, child command output is also tee'd to the log.\n" + Colors::reset);
    }

    void displayClearData()
    {
        printLine(std::string(Colors::yellow) + "--clear truncates the configured log file when save_log=true.\n" + Colors::reset);
    }

    void displayCreditsData()
    {
        printLine(std::string(Colors::yellow) + "Credits:\n" + Colors::reset);
        printLine(std::string(Colors::green) + "Developed by: Ewen MacCulloch\n" + Colors::reset);
        printLine(std::string(Colors::green) + "GitHub: Voyrox\n\n" + Colors::reset);
    }

    void displayVersionData()
    {
        const std::string currentVersion = "v1.3.1";
        printLine(std::string(Colors::green) + "  Current version: " + currentVersion + Colors::white + "\n" + Colors::reset);
        printLine(std::string(Colors::white) + "  https://github.com/Voyrox/Zippy/releases/latest\n" + Colors::reset);
    }

} // namespace commands
